tools/yarn-cling/lib/hoisting.ts

```diff
@@ -0,0 +1,51 @@
+import { PackageLockPackage } from "./types";
+
+/**
+ * Hoist package-lock dependencies in-place
+ */
+export function hoistDependencies(packageLockDeps: Record<string, PackageLockPackage>) {
+  let didChange;
+  do {
+    didChange = false;
+    simplify(packageLockDeps);
+  } while (didChange);
```

Comment:
This seems strange. Both recursion and a while loop? Pretty sure you could skip the while:
```suggestion
simplify(packageLockDeps)
```

Reply:
No, I had that at first, and it wasn't good enough. It was too hard for me to come up with the correct one-shot algorithm, It's simplest to just do multiple rounds until we can't reduce the tree anymore :).